### DISCUSSION TITLE
Missing debian-bullseye-1_amd64.deb on 2.2.0 version ?

### DISCUSSION THREAD

**Discussion by edinsof** [savonet/liquidsoap#3273] [Installation] [answered]
file is missing on release?  liquidsoap-dbgsym_2.2.0-debian-bullseye-1_amd64.deb 

which file should i run instead?

**Comment by DroidU** [comment]
We miss it too.
I tried the opam installation, but it stops with the following error message:
```
#=== ERROR while compiling ffmpeg-avcodec.1.1.8 ===============================#
# context     2.0.3 | linux/x86_64 | ocaml-base-compiler.5.0.0 | https://opam.ocaml.org#6b7cb3b3
# path        ~/.opam/5.0.0/.opam-switch/build/ffmpeg-avcodec.1.1.8
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ffmpeg-avcodec -j 8 @install
# exit-code   1
# env-file    ~/.opam/log/ffmpeg-avcodec-22047-dcc8b5.env
# output-file ~/.opam/log/ffmpeg-avcodec-22047-dcc8b5.out
### output ###
# [...]
# 16 |  (targets c_flags c_flags.sexp c_library_flags.sexp)
# 17 |  (action
# 18 |   (run ./config/discover.exe)))
# (cd _build/default/avcodec && config/discover.exe)
# which: pkg-config
# -> found: /usr/bin/pkg-config
# run: /usr/bin/pkg-config --print-errors 'libavcodec >= 58.87.100'
# -> process exited with code 1
# -> stdout:
# -> stderr:
#  | Requested 'libavcodec >= 58.87.100' but version of libavcodec is 58.35.100
# Fatal error: exception Failure("Requested 'libavcodec >= 58.87.100' but version of libavcodec is 58.35.100\n")



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build ffmpeg-avcodec 1.1.8
└─ 

```
But the situation is no better with Debian 11:
```
[ERROR] The compilation of liquidsoap-core failed at "/opt/liquidsoap/.opam/opam-init/hooks/sandbox.sh build dune build -p liquidsoap-core -j 11 --promote-install-files=false @install".

#=== ERROR while compiling liquidsoap-core.2.2.0 ==============================#
# context     2.0.8 | linux/x86_64 | ocaml-base-compiler.5.0.0 | https://opam.ocaml.org#6b7cb3b3
# path        ~/.opam/5.0.0/.opam-switch/build/liquidsoap-core.2.2.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p liquidsoap-core -j 11 --promote-install-files=false @install
# exit-code   1
# env-file    ~/.opam/log/liquidsoap-core-20045-95c722.env
# output-file ~/.opam/log/liquidsoap-core-20045-95c722.out
### output ###
# (cd _build/default && /opt/liquidsoap/.opam/5.0.0/bin/ocamlc.opt -w -40 -g -bin-annot -I src/core/.liquidsoap_ogg_flac.objs/byte -I /opt/liquidsoap/.opam/5.0.0/lib/angstrom -I /opt/liquidsoap/.opam/5.0.0/lib/bigstringaf -I /opt/liquidsoap/.opam/5.0.0/lib/bytes -I /opt/liquidsoap/.opam/5.0.0/lib/camlp-streams -I /opt/liquidsoap/.opam/5.0.0/lib/camomile/lib -I /opt/liquidsoap/.opam/5.0.0/lib/cr[...]
# File "src/core/decoder/ogg_flac_duration.ml", line 47, characters 24-54:
# 47 |         let callbacks = Flac_ogg.Decoder.get_callbacks os (fun _ -> ()) in
#                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This function has type
#          Flac.Decoder.write -> Flac_ogg.Decoder.ogg Flac.Decoder.callbacks
#        It is applied to too many arguments; maybe you forgot a `;'.



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+- The following actions failed
| - build liquidsoap-core 2.2.0
+- 
```

**Reply by vitoyucepi** [reply]
Hi @DroidU,

It seems that

- *liquidsoap=2.2.0* requires *ffmpeg=1.1.8*
- *ffmpeg=1.1.8* requires *libavcodec >= 58.87.100*
- *debian:10* provides *58.35.100*

I think if you have to provide reasonable level of support for the most recent versions of liquidsoap on not so modern Debian, then you should utilize containers or pre-built VM images.

**Comment by DigiBC** [comment]
@DroidU:

The first issue seems to be related to an outdated version of ffmpeg installed on your system.

But I also ran into the second issue when I tried to build with Dune. (OS: Linux Mint 21.2 [Ubuntu 22.04 'Jammy'] amd64)
```
File "src/core/decoder/ogg_flac_duration.ml", line 47, characters 24-54:
47 |         let callbacks = Flac_ogg.Decoder.get_callbacks os (fun _ -> ()) in
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: This function has type
         Flac.Decoder.write -> Flac_ogg.Decoder.ogg Flac.Decoder.callbacks
       It is applied to too many arguments; maybe you forgot a `;'.
```
This issue could be related to the fact that some of the released OCaml libraries in the OPAM repository are older than the development versions used to build the official packages, so that no errors were noticed there; for example:
- flac.ogg - dev version 0.5.0 instead of released version 0.4.0
- ffmpeg-av - dev version 1.1.9 instead of released version 1.1.8
- pulseaudio - dev version 0.1.6 instead of released version 0.1.5
- theora - dev version 0.4.1 instead of released version 0.4.0
- vorbis - dev version 0.8.1 instead of released version 0.8.0

Finally, I ran into an issue with the gstreamer library.
```
File "src/core/decoder/gstreamer_decoder.ml", line 173, characters 3-27:
173 |   ({ Decoder.decode; seek }, close, gst.bin)
         ^^^^^^^^^^^^^^^^^^^^^^^^
Error: Some record fields are undefined: eof
File "src/core/encoder/encoders/gstreamer_encoder.ml", line 208, characters 13-14:
208 |         Some f
                   ^
Error: This expression has type
         'a -> Meta_format.export_metadata -> Encoder.encoder
       but an expression was expected of type
         Encoder.factory =
           pos:Pos.t option ->
           string -> Meta_format.export_metadata -> Encoder.encoder
```
I'm aware that gstreamer is rated "deprecated" now, but it's still an available optional input / output.

Let's hope that updated OCaml libraries will be released soon...

**Reply by vitoyucepi** [reply]
@toots, I can confirm this problem with gstreamer.
I can't build liquidsoap *2.2.0* with gstreamer support from opam.
`opam install liquidsoap=2.2.0 gstreamer`.

@DigiBC, I'm trying to reproduce the problem with flac.
What commands are you using to install opam, ocaml and liquidsoap?
For example I'm using
- `opam --version` - `2.1.5`
- `opam switch` - `4.14.1`
- ```sh
  opam init --bare --disable-sandboxing
  opam switch create 4.14.1
  opam install liquidsoap flac
  ```

**Reply by DigiBC** [reply]
@vitoyucepi:

Originally, I'd tried a different installation approach. (I'm currently trying to build my own DEB packages - but that's another story with other challenges...)

So, I tested your exemplary way on a different system: As expected, the error is exactly the same.

The `ocaml-flac` library in the official opam package repository is still version 0.4.0, but Liquidsoap version 2.2.0 requires version 0.5.0. Apparently, adding the updated version of `ocaml-flac` to the repository was overlooked when the new liquidsoap version was released. I hope it will be updated soon...

For the time being, I have found a **workaround**:
Liquidoap's help pages recommend using [opam pinning](https://www.liquidsoap.info/doc-2.2.0/build.html#opam-pinning) to update a dependency to its latest version.

`opam pin -ny git+https://github.com/savonet/ocaml-flac`

That did the trick!   :smiley:

**Reply by vitoyucepi** [reply]
> Liquidsoap version 2.2.0 requires version 0.5.0

Are you sure?
As I can see on the opam web page https://opam.ocaml.org/packages/liquidsoap-core/liquidsoap-core.2.2.0/ liquidsoap *2.2.0* conflicts with flac < *0.3.0*.
You can see the same here. https://github.com/savonet/liquidsoap/blob/6eba12641e468c435114b2afc8763211f41b168a/dune-project#L122

```
2023/07/29 15:39:46 [main:3] Liquidsoap 2.2.0+dev
2023/07/29 15:39:46 [main:3] Using: angstrom=0.15.0 bigarray=[distributed with OCaml] bigstringaf=0.9.1 bytes=[distributed with OCaml] camlp-streams camomile.lib=2.0 cry=1.0.1 curl=0.9.2 dtools=0.4.5 dune-build-info=3.9.2 dune-private-libs.dune-section=3.9.2 dune-site=3.9.2 dune-site.private=3.9.2 duppy=0.9.3 fileutils=0.6.4 flac=0.4.0 gen=1.1 liquidsoap-lang=2.2.0 liquidsoap-lang.console=2.2.0 liquidsoap_builtins liquidsoap_core liquidsoap_flac liquidsoap_optionals liquidsoap_oss liquidsoap_runtime menhirLib=20230608 metadata=0.2.0 mm=0.8.4 mm.audio=0.8.4 mm.base=0.8.4 mm.image=0.8.4 mm.midi=0.8.4 mm.video=0.8.4 pcre=7.5.0 sedlex=3.2 seq=[distributed with OCaml 4.07 or above] stdlib-shims=0.3.0 str=[distributed with OCaml] stringext=1.6.0 threads=[distributed with OCaml] threads.posix=[distributed with OCaml] unix=[distributed with OCaml] uri=4.2.0
```
```
flac                  0.4.0       Bindings to libflac
liquidsoap            2.2.0       Swiss-army knife for multimedia streaming
```

**Reply by DigiBC** [reply]
Yes, I'm sure!
According to my simple mathematical understanding version 0.5.0 is _higher_ and _not lower_ than version 0.3.0, so there is no conflict.   :smirk:
Just try it out...!

**Reply by vitoyucepi** [reply]
Conflicts with flac < *0.3.0* means it requires flac >= *0.3.0*.
These flac package versions feat:
- *0.3.0*
- *0.3.1*
- *0.4.0*

I've installed the oldest suitable version using the command `opam install liquidsoap=2.2.0 flac=0.3.0`.

**Reply by DigiBC** [reply]
Anyway, if you take a look at the configuration files associated with each of the released packages - e.g. [liquidsoap_2.2.0-ubuntu-jammy-1.config](https://github.com/savonet/liquidsoap-release-assets/releases/download/v2.2.0/liquidsoap_2.2.0-ubuntu-jammy-1.config) - you'll find under the "Libs versions" section: `flac=0.5.0`

**Reply by vitoyucepi** [reply]
Aha! Certainly, these are not requirements but build dependencies managed in the [savonet/liquidsoap-full](https://github.com/savonet/liquidsoap-full) repository. To stay in sync with the liquidsoap release cycle, you have to pull+pin the exact commits.

Generally, you don't have to pin commits, but rather do `opam install`. Some packages from opam require additional interactions. In that situation, you will need to conduct research. For example, I tried to install srt using opam to debug this problem and described my experience in savonet/ocaml-srt#4.

**Reply by martinkirch** [reply]
Hello,
I can confirm the 2.2.x rolling release branch needs ocaml-flac-0.5.0 (with ocam5) - I had the same error as https://github.com/savonet/liquidsoap/discussions/3273#discussioncomment-6577154 and fixed it with:
```
git clone https://github.com/savonet/ocml-flac.git
cd ocaml-flac/
opam install .
```
then back to Liquidsoap and `opam install .`

However, now I'm facing a segfault when starting the radio.

**Reply by vitoyucepi** [reply]
Hi @martinkirch,
Sorry, but I failed to replicate this behavior.

I tested in container `debian:trixie-slim` with the following script
```
apt-get update
apt-get dist-upgrade -y
apt-get install --no-install-recommends -y curl ca-certificates;
bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)"
apt-get install --no-install-recommends -y patch unzip bubblewrap git bzip2;
apt-get install --no-install-recommends -y gcc make libc6-dev;
opam init --bare --disable-sandboxing
opam switch create 5.0.0
eval $(opam env --switch=5.0.0)
git clone https://github.com/savonet/liquidsoap.git
cd liquidsoap
git checkout rolling-release-v2.2.x
opam pin -n .
opam install liquidsoap flac
```
Did I miss something?

**Reply by martinkirch** [reply]
Hello,
Your test looks good. Which version of `ocaml-flac` does that install ?
I ran it in debian bookworm, maybe it causes a difference

**Reply by vitoyucepi** [reply]
Everything works fine, except [port binding](https://github.com/savonet/liquidsoap/issues/3382).

<details>
<summary>Script</summary>

1. `docker run -it --rm --network host debian:bookworm-slim bash`
1. ```sh
   apt-get update;
   apt-get dist-upgrade -y;
   apt-get install --no-install-recommends -y curl ca-certificates patch unzip bubblewrap git bzip2 gcc make libc6-dev;
   bash -c "sh <(curl -fsSL https://raw.githubusercontent.com/ocaml/opam/master/shell/install.sh)";
   opam init --bare --disable-sandboxing;
   opam switch create 5.0.0;
   eval $(opam env --switch=5.0.0);
   git clone https://github.com/savonet/liquidsoap.git;
   cd liquidsoap;
   git checkout rolling-release-v2.2.x;
   opam pin -n .;
   opam install liquidsoap flac; 
   ```
   The same for *4.14.1*.
1. `1.liq`
   ```ruby
   output.harbor(%flac, sine(), mount='/test')
   ```
1. ```
   flac                  0.4.0       Bindings to libflac
   liquidsoap            2.2.1       pinned to version 2.2.1 at git+file:///liquidsoap#HEAD
   ocaml                 5.0.0       The OCaml compiler (virtual package)
   ```

</details>

<details>
<summary>Log</summary>

```
2023/09/03 13:35:52 >>> LOG START
2023/09/03 13:35:51 [main:3] Liquidsoap rolling-release-v2.2.x+git@46968a5a
2023/09/03 13:35:51 [main:3] Using: angstrom=0.15.0 bigarray=[distributed with OCaml] bigstringaf=0.9.1 bytes=[distributed with OCaml] camlp-streams camomile.lib=2.0 cry=1.0.1 curl=0.9.2 dtools=0.4.5 dune-build-info=3.10.0 dune-private-libs.dune-section=3.10.0 dune-site=3.10.0 dune-site.private=3.10.0 duppy=0.9.3 fileutils=0.6.4 flac=0.4.0 gen=1.1 liquidsoap-lang=rolling-release-v2.2.x liquidsoap-lang.console=rolling-release-v2.2.x liquidsoap_builtins=rolling-release-v2.2.x-dirty liquidsoap_core=rolling-release-v2.2.x-dirty liquidsoap_flac=rolling-release-v2.2.x-dirty liquidsoap_optionals=rolling-release-v2.2.x-dirty liquidsoap_oss=rolling-release-v2.2.x-dirty liquidsoap_runtime=rolling-release-v2.2.x-dirty menhirLib=20230608 metadata=0.2.0 mm=0.8.4 mm.audio=0.8.4 mm.base=0.8.4 mm.image=0.8.4 mm.midi=0.8.4 mm.video=0.8.4 pcre=7.5.0 sedlex=3.2 seq=[distributed with OCaml 4.07 or above] stdlib-shims=0.3.0 str=[distributed with OCaml] stringext=1.6.0 threads=[distributed with OCaml] threads.posix=[distributed with OCaml] unix=[distributed with OCaml] uri=4.2.0
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [main:3] DISCLAIMER: This version of Liquidsoap has been compiled from a snapshot of the
2023/09/03 13:35:51 [main:3] development code. As such, it should not be used in production unless you know
2023/09/03 13:35:51 [main:3] what you are doing!
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [main:3] We are, however, very interested in any feedback about our development code and
2023/09/03 13:35:51 [main:3] committed to fix issues as soon as possible.
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [main:3] If you are interested in collaborating to the development of Liquidsoap, feel
2023/09/03 13:35:51 [main:3] free to drop us a mail at <savonet-devl@lists.sf.net> or to join the slack chat
2023/09/03 13:35:51 [main:3] at <http://slack.liquidsoap.info>.
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [main:3] Please send any bug report or feature request at
2023/09/03 13:35:51 [main:3] <https://github.com/savonet/liquidsoap/issues>.
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [main:3] We hope you enjoy this snapshot build of Liquidsoap!
2023/09/03 13:35:51 [main:3] 
2023/09/03 13:35:51 [clock:3] Using builtin (low-precision) implementation for latency control
2023/09/03 13:35:52 [main:3] Standard library loaded in 0.70 seconds.
2023/09/03 13:35:52 [frame:3] Using 44100Hz audio, 25Hz video, 44100Hz main.
2023/09/03 13:35:52 [frame:3] Video frame size set to: 1280x720
2023/09/03 13:35:52 [frame:3] Frame size must be a multiple of 1764 ticks = 1764 audio samples = 1 video samples.
2023/09/03 13:35:52 [frame:3] Targeting 'frame.duration': 0.04s = 1764 audio samples = 1764 ticks.
2023/09/03 13:35:52 [frame:3] Frames last 0.04s = 1764 audio samples = 1 video samples = 1764 ticks.
2023/09/03 13:35:52 [sandbox:3] Sandboxing disabled
2023/09/03 13:35:52 [startup:3] Typechecking: 0.48s
2023/09/03 13:35:52 [startup:3] Evaluation: 0.01s
2023/09/03 13:35:52 [startup:3] Typechecking: 0.04s
2023/09/03 13:35:52 [startup:3] Evaluation: 0.00s
2023/09/03 13:35:52 [startup:3] Typechecking: 0.00s
2023/09/03 13:35:52 [startup:3] Evaluation: 0.00s
2023/09/03 13:35:52 [startup:3] Loaded 1.liq: 0.00s
2023/09/03 13:35:52 [video.converter:3] Couldn't find preferred video converter: ffmpeg.
2023/09/03 13:35:52 [audio.converter:3] Using samplerate converter: native.
2023/09/03 13:35:52 [/test:3] Content type is {audio=pcm(stereo)}.
2023/09/03 13:35:52 [sine:3] Content type is {audio=pcm(stereo)}.
2023/09/03 13:35:52 [harbor:3] Adding handler for 'GET ^/test$' on port 8000
2023/09/03 13:35:52 [clock.main:3] Streaming loop starts in auto-sync mode
2023/09/03 13:35:52 [clock.main:3] Delegating synchronization to CPU clock
2023/09/03 13:35:52 [video.text:3] Using native implementation
^C2023/09/03 13:35:58 [main:3] Shutdown signal received.
2023/09/03 13:35:58 [threads:3] Main loop exited
2023/09/03 13:35:58 [main:3] Shutdown started!
2023/09/03 13:35:58 [threads:3] Waiting for main threads to terminate...
2023/09/03 13:35:58 [clock.main:3] Streaming loop stopped.
2023/09/03 13:35:58 [threads:3] Main threads terminated.
2023/09/03 13:35:58 [threads:3] Shutting down scheduler...
```

</details>

**Reply by martinkirch** [reply]
@vitoyucepi I just noticed that the compile error involves `Flac_ogg` ; can you add `opam install ogg` to your test container ?

**Comment by DroidU** [comment]
Our goal would be to be able to use the same version of Liquidsoap from Debian 10 to 12. Reason for the same versions: The streaming control panel we developed is supported from Debian 10 to 12, and many of our customers are still using Debian 10. We also need Stereo Tool support.

**Reply by toots** [reply]
@DroidU you might need to look for a backport of ffmpeg. This is a complex library with various dependencies that might, in turn, also need backport or to be dropped from your build.

Another alternative is to use a virtualized container. We do provide docker images as well.

**Comment by toots** [comment]
@edinsof  I'm not sure why the automated debian build skipped that file. It might not have been produced. This package is totally ooptional so you can just ignore it.

**Reply by vitoyucepi** [reply, marked as answer]
@toots, *2.2.0* release has no builds for debian bullseye amd64.

**Reply by toots** [reply]
Well that explains heh. I got confused b/c there's only one for armhf.

The reason we have one for armhf is because the most likely users of this build are raspberry pi, which are still running this debian.

Otherwise, due to limited resources, we always maintain current stable + testing for debian and the release happened a couple of weeks ago..

For bullseye users, I would suggest to use a docker image.
